### PR TITLE
Joy teleop revamp

### PR DESCRIPTION
**Pull request by clalancette**
This is a complete revamp of the joy_teleop code.  The main reasons to do this are:

1.  Readability.  This version is much easier to read and reason about.  Which leads to...
1.  Fix some bugs in the implementation, particularly with different things acting differently.
1.  Add in more error checking to give the user early feedback that their config may not work.
1.  Also add in tests for all of the major functionality.

Note that the tests will only work on Eloquent or later, since they depend on the launch_testing framework that isn't available until Eloquent.  Thus, I'd suggest that we create an `eloquent-devel` branch for this to be merged into.

This is somewhat related to #46 .  We can either merge #46 and then I can rebase this, or we can just close #46 in preference to this one.  Either is fine with me.

Questions, comments, reviews are all appreciated!

### CODE REVIEW

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (hidmic) [inline joy_teleop/joy_teleop/joy_teleop.py]: @clalancette nit^N: there's [`str.partition`](https://docs.python.org/3.7/library/stdtypes.html#str.partition) too

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (hidmic) [inline joy_teleop/joy_teleop/joy_teleop.py]: @clalancette won't this result in the same error log being printed twice (once here, once when this exception gets caught in `main`)? Same elsewhere.

File: joy_teleop/test/test_action_fibonacci.py
Comment (hidmic) [inline joy_teleop/test/test_action_fibonacci.py]: @clalancette nit: here and elsewhere, can we use fully qualified QoS from `rclpy.qos`?

File: joy_teleop/test/test_action_fibonacci.py
Comment (hidmic) [inline joy_teleop/test/test_action_fibonacci.py]: @clalancette why mixing both styles?

File: joy_teleop/test/test_action_fibonacci.py
Comment (hidmic) [inline joy_teleop/test/test_action_fibonacci.py]: @clalancette nit: here and elsewhere, this may never happen if an previous assertion fails.

File: joy_teleop/test/test_action_fibonacci.py
Comment (hidmic) [inline joy_teleop/test/test_action_fibonacci.py]: @clalancette nit: here and elsewhere, both `launch_service` and `proc_output` seem unused.

File: joy_teleop/test/test_action_fibonacci.py
Comment (hidmic) [inline joy_teleop/test/test_action_fibonacci.py]: @clalancette meta: here and elsewhere, one thing you're completely allowed to do is to drop common code for all tests into a separate module and reuse (e.g. having a common function to get the test launch description or having a custom 
 unittest.TestCase` subclass)

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (clalancette) [reply joy_teleop/joy_teleop/joy_teleop.py]: Simple boolean.  I'll update it.

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (clalancette) [reply joy_teleop/joy_teleop/joy_teleop.py]: Hm, that's a good question.  I originally thought that this was dynamic (like the topic 'axis_mappings'), but it really isn't.  We could just do this once in the constructor and reuse that.  I'll fix that up.

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (clalancette) [reply joy_teleop/joy_teleop/joy_teleop.py]: I tried this, and it turns out to be exactly the same lines of code.  Since I don't expect the list of interface_groups to expand much over time, and since I find this more readable, I left it as-is.

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (clalancette) [reply joy_teleop/joy_teleop/joy_teleop.py]: I'm not exactly sure what you are thinking about here.  I guess we could make the assumption that if there are no dots in the parameter, it can't possibly be for us, but I think that's the only heuristic I would apply.  Other ones might have false positives.  Do you have other concrete suggestions?

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (clalancette) [reply joy_teleop/joy_teleop/joy_teleop.py]: Ah, good call.  I forgot about the exception handling in main.  Will remove the redundancy.

File: joy_teleop/test/test_action_fibonacci.py
Comment (clalancette) [reply joy_teleop/test/test_action_fibonacci.py]: Oh, good question.  I wrote the code at different times, and I just didn't realize.  Which one is preferred?

File: joy_teleop/test/test_action_fibonacci.py
Comment (clalancette) [reply joy_teleop/test/test_action_fibonacci.py]: For that matter, `joy_teleop_node` isn't used either.  Will simplify the signatures in all of the tests.

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (clalancette) [reply joy_teleop/joy_teleop/joy_teleop.py]: Hm, true.  I don't know that one is more readable than the other in this case.  Is there an additional benefit to using `str.partition`?

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (hidmic) [reply joy_teleop/joy_teleop/joy_teleop.py]: I presume a `TypeError` exception may get thrown because we blindly attempt to use every parameter value as a `dict`. If we only care about those that represent a `joy_teleop` config, why not ignoring the rest? I'd think we can check and drop them after looping in `retrieve_config()`. WDYT?

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (hidmic) [reply joy_teleop/joy_teleop/joy_teleop.py]: I'd say that API is there for exactly this use case, but otherwise both should be equally performant (unless a `tuple` is objectively cheaper to instantiate than a `list`).

File: joy_teleop/test/test_action_fibonacci.py
Comment (hidmic) [reply joy_teleop/test/test_action_fibonacci.py]: In `unittest`-style tests, IIRC `self.assertTrue` is preferred.

Comment (clalancette) [reply joy_teleop/test/test_action_fibonacci.py]: Ah yes, thanks.  I just read the docs, and the difference is that unittest collects failures for `assertTrue` and reports all of them, while `assert` just terminates immediately.  So yeah, I think `assertTrue` is better; will switch to that.

File: joy_teleop/test/test_action_fibonacci.py
Comment (clalancette) [reply joy_teleop/test/test_action_fibonacci.py]: Hm, true.

The truly correct way to do this would be something like:

```
pub = create_publisher()
try:
    timer = create_timer()
    try:
        action_server = ActionServer()
        try:
          # Do test stuff in here
        finally:
          action_server.destroy()
    finally:
        destroy_timer(timer)
finally:
    pub.destroy()
```

But it is pretty wordy.  Other thoughts on how to do this nicely?

Comment (clalancette) [reply joy_teleop/test/test_action_fibonacci.py]: Never mind this one, actually.  After the other revamps you suggested, this became a lot easier.

**Comment by clalancette** [comment]
@hidmic Thanks for the initial review!  I've addressed all of your comments in the latest code.  Note that I ended up adding the changes to the respective commits, since I want the commits to be self-contained and build off of each other.  Thus, this whole thing is rebased and modified.  Please take another look when you get a chance.

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (hidmic) [inline joy_teleop/joy_teleop/joy_teleop.py]: @clalancette should this be 
```suggestion
        self.active = False
```
?

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (hidmic) [inline joy_teleop/joy_teleop/joy_teleop.py]: @clalancette should this be
```suggestion
        if not self.active:
```
?

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (hidmic) [inline joy_teleop/joy_teleop/joy_teleop.py]: @clalancette I'd say the comment about not making any noise when something expected comes you way still stands, but it's not a strong stand.

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (hidmic) [reply joy_teleop/joy_teleop/joy_teleop.py]: Same about this applying, but won't block on it.

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (clalancette) [reply joy_teleop/joy_teleop/joy_teleop.py]: Yep, fixed in the latest.

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (clalancette) [reply joy_teleop/joy_teleop/joy_teleop.py]: Fixed in the latest.

File: joy_teleop/joy_teleop/joy_teleop.py
Comment (clalancette) [reply joy_teleop/joy_teleop/joy_teleop.py]: I was kind of leaving this in since it was already there, and it is a warning to the user if they entered in something that they didn't mean.  But I think we have enough other error checking now that it isn't needed as much.  So I reduced this even further to a `debug`, so you won't see it by default.

**Comment by clalancette** [comment]
> Same about this applying, but won't block on it.

I ended up switching to `str.partition`; it's the same number of lines of code, but slightly nicer since we don't have to specify a stop number of splits.

Thanks again for the review!

**Comment by clalancette** [comment]
Any thoughts on this revamp from the maintainers?  I'd like to try to get this landed and released for Foxy if at all possible (~ 1 month away).  Thanks!

**Comment by bmagyar** [comment]
@clalancette thanks very much for this and the patience as well, I am currently a little overwhelmed with the amount of stuff coming on the `ros_control` front while this package was relatively low bandwidth :) I'm happy to add a new maintainer if you guys know a volunteer *wink wink*

I've merged #46 to dashing as I think it's great to have it asap. I've pushed a release of it to dashing and created an eloquent branch to which I've retargeted this MR to. Please resolve the conflict now.

**Comment by bmagyar** [comment]
@clalancette I know that refactoring work comes with touching a lot of code but I'd like to ask you to try aiming for smaller changesets in the future if possible. Around 500 lines is an amount that no reviewer will be intimidated to do while this current 1600 is very prone to get to the backlog. Not all ROS2 community members are from companies, some of us actually do it in our free time :)

**Comment by clalancette** [comment]
@bmagyar Thanks for taking a look, it is appreciated.  I've now rebased this onto eloquent-devel and verified that the tests still pass.

Understood about large PRs; I know they are hard to review.  As you say, this is a large revamp so it grew large.  I will point out that each commit is a self-contained change, so reviewing this commit-by-commit should be pretty straightforward.  That being said, if you'd like me to split this up into 3 different PRs (one to do minor cleanup, one to refactor, and one to add the tests), I can do that.  Just let me know.  And thanks again!

**Comment by bmagyar** [comment]
@clalancette I have to say I quite like the PR but would still like some more time to properly go through it. Also the next time we meet I'll pass you a lot of cookies for the tests on this PR.

Is there any chance someone could backport the tests to Noetic? I'm currently working on figuring out what's missing for that release and tests would be awesome to have in the future. ( Too late for this release though)

**Comment by clalancette** [comment]
> @clalancette I have to say I quite like the PR but would still like some more time to properly go through it. Also the next time we meet I'll pass you a lot of cookies for the tests on this PR.

Great!  Thanks for taking a look at it.

> Is there any chance someone could backport the tests to Noetic? I'm currently working on figuring out what's missing for that release and tests would be awesome to have in the future. ( Too late for this release though)

Unfortunately, not that easily.  The project I did this for originally only cares about ROS 2, and if I'm honest, I don't really know much about testing for ROS 1.

**Comment by clalancette** [comment]
Friendly ping here, when you have time @bmagyar .

**Comment by clalancette** [comment]
Another friendly ping on this one to review and get in.  I'd like to then get this released into Foxy and Rolling.

**Comment by bmagyar** [comment]
Thanks for the many pings, I'll try to fit this into today.

File: joy_teleop/test/test_get_interface_type.py
Comment (bmagyar) [inline joy_teleop/test/test_get_interface_type.py:18]: You may want to update this

File: joy_teleop/test/test_parameter_failures.py
Comment (bmagyar) [inline joy_teleop/test/test_parameter_failures.py:18]: same here

File: joy_teleop/test/test_service_add_two_ints.py
Comment (bmagyar) [inline joy_teleop/test/test_service_add_two_ints.py:18]: as well

File: joy_teleop/test/test_service_not_ready.py
Comment (bmagyar) [inline joy_teleop/test/test_service_not_ready.py:18]: .

File: joy_teleop/test/test_service_trigger.py
Comment (bmagyar) [inline joy_teleop/test/test_service_trigger.py:18]: pal again

File: joy_teleop/test/test_topic_axis_mappings.py
Comment (bmagyar) [inline joy_teleop/test/test_topic_axis_mappings.py:18]: .

File: joy_teleop/test/test_topic_message_value.py
Comment (bmagyar) [inline joy_teleop/test/test_topic_message_value.py:18]: .

**Review by bmagyar — CHANGES_REQUESTED** [review]
@clalancette everythin ggood, many thanks to @hidmic for previous reviews, I only found some license typos

**Comment by bmagyar** [comment]
@clalancette I know this is a tricky question and definitely out of scope for this PR but would you be open to add typing to this codebase? The tests are already a great addition, types would further enhance maintainability

File: joy_teleop/test/test_get_interface_type.py
Comment (clalancette) [reply joy_teleop/test/test_get_interface_type.py]: Fixed now.

File: joy_teleop/test/test_parameter_failures.py
Comment (clalancette) [reply joy_teleop/test/test_parameter_failures.py]: Fixed.

File: joy_teleop/test/test_service_add_two_ints.py
Comment (clalancette) [reply joy_teleop/test/test_service_add_two_ints.py]: Fixed.

File: joy_teleop/test/test_service_trigger.py
Comment (clalancette) [reply joy_teleop/test/test_service_trigger.py]: Fixed.

File: joy_teleop/test/test_service_not_ready.py
Comment (clalancette) [reply joy_teleop/test/test_service_not_ready.py]: Fixed.

File: joy_teleop/test/test_topic_axis_mappings.py
Comment (clalancette) [reply joy_teleop/test/test_topic_axis_mappings.py]: Fixed.

File: joy_teleop/test/test_topic_message_value.py
Comment (clalancette) [reply joy_teleop/test/test_topic_message_value.py]: Fixed.

**Comment by clalancette** [comment]
@bmagyar Thanks so much for doing the review!

> @clalancette I know this is a tricky question and definitely out of scope for this PR but would you be open to add typing to this codebase? The tests are already a great addition, types would further enhance maintainability

Yep, I ended up doing this in c8903b2.  It actually uncovered a couple of bugs too, which is a good thing.  Please take another look, thanks.

**Review by bmagyar — APPROVED** [review]
Thanks so much for this @clalancette !

**Comment by clalancette** [comment]
> Thanks so much for this @clalancette !

Perfect, thank you!  If I could be so bold and ask for a release into Eloquent, Foxy, and Rolling, that would be great.

**Pull request merged**
